File: 383.ransom-note.py
#
# @lc app=leetcode id=383 lang=python3
#
# [383] Ransom Note
#

# @lc code=start
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        word_count = {}
        for c in magazine:
            if not c in word_count:
                word_count[c] = 0
            word_count[c] += 1
        for c in ransomNote:
            if not c in word_count or not word_count[c]:
                return False
            word_count[c] -= 1
        return True
        
# @lc code=end

File: 383.ransom-note.go
/*
 * @lc app=leetcode id=383 lang=golang
 *
 * [383] Ransom Note
 */
package main

// @lc code=start
func canConstruct(ransomNote string, magazine string) bool {

	m := map[byte]int{}
	for i := 0; i < len(magazine); i++ {
		char := magazine[i]
		m[char] += 1
	}

	for i := 0; i < len(ransomNote); i++ {
		char := ransomNote[i]
		count, ok := m[char]
		if !ok {
			return false
		}
		if count == 0 {
			return false
		}
		m[char] -= 1
	}
	return true
}

// @lc code=end
